Services/IotHub/IotHubConnectionStringManager.cs

```diff
@@ -118,33 +122,122 @@ public string RedactAndStore(string connectionString)
+            catch (Exception e)
+            {
+                string message = "Could not read devices with IotHub connection " +
+                                 "string provided. Check that the policy for the key allows " +
+                                 "`Registry Read/Write` and `Service Connect` permissions.";
+                this.log.Debug(message, () => { });
+                throw new IotHubConnectionException(message, e);
+            }
+
+        }
+
+        private async Task ValidateWritePermissionsAsync(string connectionString)
+        {
+            var registryManager = RegistryManager.CreateFromConnectionString(connectionString);
+
+            string testDeviceId = "test-device-creation" + DateTime.UtcNow.ToFileTimeUtc();
+            var device = new Device(testDeviceId);
+
+            // To test permissions, try to create a test device.
+            try
+            {
+                await registryManager.AddDeviceAsync(device);
+            }
+            catch (Exception e)
+            {
+                string message = "Could not create devices with IotHub connection " +
+                                 "string provided. Check that the policy for the key allows " +
+                                 "`Registry Read/Write` and `Service Connect` permissions.";
+                this.log.Debug(message, () => { });
+                throw new IotHubConnectionException(message, e);
+            }
+
+            // Delete the test device that was created.
+            // If test device deletion fails, retry. Throw if unsuccessful.
+            int deleteRetryCount = 3;
+            Device response;
+            do
+            {
+                deleteRetryCount--;
+
                 try
                 {
-                    RegistryManager.CreateFromConnectionString(connectionString);
+                    await registryManager.RemoveDeviceAsync(testDeviceId);
                 }
-                catch (IOException)
+                catch (Exception e)
                 {
-                    string message = "Could not connect to IotHub with the connection" +
-                                     "string provided. Check that the key is valid and " +
-                                     "that the hub exists.";
-                    throw new IotHubConnectionException(message);
+                    string message = "Could not delete test device from IotHub. Device Id: " + testDeviceId;
+                    this.log.Error(message, () => new { testDeviceId });
+                    throw new IotHubConnectionException(message, e);
                 }
+
+                response = await registryManager.GetDeviceAsync(testDeviceId);
+
+            } while (response != null && deleteRetryCount > 0);
+
+            if (response != null)
+            {
+                string message = "Could not delete test device from IotHub. Device Id: " + testDeviceId;
+                this.log.Error(message, () => new { testDeviceId });
```

Comment:
log also the number of attempts.  No need to include the device id in the msg (avoid string manipulation and reduce GC work)